pkg/kgo/pools.go

```diff
@@ -0,0 +1,207 @@
+package kgo
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/twmb/franz-go/pkg/kmsg"
+)
+
+////////////////////////////////////////////////////////////////
+// NOTE:                                                      //
+// NOTE: Make sure new hooks are checked in implementsAnyPool //
+// NOTE:                                                      //
+////////////////////////////////////////////////////////////////
+
+// Pool is a memory pool to be used where relevant.
+//
+// The base Pool interface is meaningless, but wherever a type can be pooled in
+// kgo, the client checks if your pool implements an appropriate pool interface.
+// If so, the pool is received from (Get), and when the data is done being used,
+// the pool is put back into (Put).
+//
+// All pool interfaces in this package have Pool in the name. Pools must be safe
+// for concurrent use.
+type Pool any
+
+type pools []Pool
+
+func (ps pools) each(fn func(Pool)) {
+	for _, p := range ps {
+		fn(p)
+	}
+}
+
+/*TODO
+type PoolRequestBuffer interface {
```

Comment:
did you mean to implement this in this PR or only later on?

Reply:
This will be implemented in this PR soon :)

Reply:
Actually, I'm reverting the implementation. There are not many ways a person can really implement this buffer pool and the benefit of user configurability here seems low at the moment. If this is requested in the future, it can be added.